### ISSUE TITLE
Ability to wrap errors

### ISSUE THREAD

**Issue by terinjokes** [bodenr/errors#9]
I want to wrap a low-level error with another error, and have `errors`'s "stack" setting to affect how it's printed.

```javascript
var fs = require('fs');
var errors = require('errors');

function doSomeImportantThing(filename, cb) {
  lowLevelFileSystemThing(filename, function(err) {
    var newErr =  new (errors.find('SomeImportantFailure'))({
      cause: err
    });

   console.log(newError.toJSON());
  });
}
```

stack is `false`:

```json
{
   "cause": {
     "message": "lowLevelFileSystemThing failed foobarred $filename"
   },
  "code": "600",
  "message": "SomeImportantThing Failed",
  "name": "SomeImportFailure"
}
```

stack is `true`:

```json
{
   "cause": {
     "message": "lowLevelFileSystemThing foobarred $filename",
     "stack":  "…"
   },
  "code": "600",
  "message": "SomeImportantThing Failed",
  "name": "SomeImportFailure",
  "stack": "…"
}
```

The current output prints the "stack" in all cases.

**Comment by bodenr** [comment]
@terinjokes with the current impl of errors you can do the following:

sample code:
```js
var errors = require('errors');

errors.create({
    name: 'WrappedError'
});

errors.create({
    name: 'UnexpectedError'
});

function raise() {
    try {
        throw new errors.WrappedError('The cause');
    } catch (e) {
        return new errors.UnexpectedError({
            cause: e.toJSON() // capture stack if enabled
        });
    }
}



console.log("Without stacks:");
console.log("-----------");
console.log(raise().toJSON());

errors.stacks(true);
console.log("\nWith stacks:");
console.log("-----------");
console.log(raise().toJSON());
```

which outputs something like:
```
Without stacks:
-----------
{ code: 602,
  status: 500,
  name: 'UnexpectedError',
  message: 'An unexpected UnexpectedError occurred.',
  cause: 
   { code: 601,
     status: 500,
     name: 'WrappedError',
     message: 'The cause' } }

With stacks:
-----------
{ stack: 'Error: An unexpected UnexpectedError occurred.\n    at raise (/tmp/errors/err.js:15:16)\n    at Object.<anonymous> (/tmp/errors/err.js:30:13)\n    at Module._compile (module.js:456:26)\n    at Object.Module._extensions..js (module.js:474:10)\n    at Module.load (module.js:356:32)\n    at Function.Module._load (module.js:312:12)\n    at Function.Module.runMain (module.js:497:10)\n    at startup (node.js:119:16)\n    at node.js:906:3',
  code: 602,
  status: 500,
  name: 'UnexpectedError',
  message: 'An unexpected UnexpectedError occurred.',
  cause: 
   { stack: 'Error: The cause\n    at raise (/tmp/errors/err.js:13:15)\n    at Object.<anonymous> (/tmp/errors/err.js:30:13)\n    at Module._compile (module.js:456:26)\n    at Object.Module._extensions..js (module.js:474:10)\n    at Module.load (module.js:356:32)\n    at Function.Module._load (module.js:312:12)\n    at Function.Module.runMain (module.js:497:10)\n    at startup (node.js:119:16)\n    at node.js:906:3',
     code: 601,
     status: 500,
     name: 'WrappedError',
     message: 'The cause' } }
```

does this functionality suffice for your use case(s)? if not please provide additional details.

thanks

**Comment by terinjokes** [comment]
There's two cases I want to handled with "cause":

* The error I'm wrapping is from my own code (just something lower-level), in which case it's already an instance of `errors`, already works, no `.toJSON` needed.
* The error I'm wrapping comes from somewhere else, either from the JavaScript runtime itself, Node's built-in modules, and other third-party modules that throw or pass errors. These are likely not instances of `errors`.

**Comment by bodenr** [comment]
@terinjokes my current thought is to export a function (called `extract` below) on the `errors` module which accepts a native exception (node, js runtime, etc.) and returns a js object which contains error's related attrs (stack, message, etc.) from the native exception. Note: the function would also accept an optional map (a default map provided for most uses) to define how attrs of the native exception map into the extracted object.

Your code above would then become something like:
```js
var fs = require('fs');
var errors = require('errors');

function doSomeImportantThing(filename, cb) {
  lowLevelFileSystemThing(filename, function(err) {
    var newErr =  new (errors.find('SomeImportantFailure'))({
      cause: errors.extract(err)
    });

   console.log(newError.toJSON());
  });
}
```

I believe this approach should suit your use case, while at the same time providing the general ability to extract errors related attrs from a native exception.

Thoughts on this approach?

**Comment by bodenr** [comment]
@terinjokes do you still need this functionality? if so please see my previous comment.

**Comment by terinjokes** [comment]
@bodenr I do, but I have not had the time to try out your suggestion: I've been working on a different project.

**Comment by bodenr** [comment]
@terinjokes - I wasn't asking you to try anything; was just verifying the proposed impl would work for you.

**Comment by terinjokes** [comment]
@bodenr Sorry, I was using "try" to mean think about it and get back to you.

**Comment by bodenr** [comment]
@terinjokes -- I pushed a prototype branch (boden/format-native-errors) which implements the behavior I described above.

See https://github.com/bodenr/errors/blob/boden/format-native-errors/lib/errors.js#L140 which implements a new errors function called ```errorToJSON``` -- see the source comments for more details.

Sample usage:
```js
var errors = require('errors');
errors.stacks(true);
errors.create({
    name: 'NativeError'
});

var te = new TypeError("bad type");

console.log("%j", new errors.NativeError("Fake native error").toJSON());
console.log("------------");
console.log("%j", errors.errorToJSON(te));
console.log("------------");
console.log("%j", new errors.NativeError({cause: errors.errorToJSON(te)}).toJSON());
```

When run produces something like:
```
{"stack":"Error: Fake native error\n    at Object.<anonymous> (/tmp/extractError.js:9:19)\n    at Module._compile (module.js:456:26)\n    at Object.Module._extensions..js (module.js:474:10)\n    at Module.load (module.js:356:32)\n    at Function.Module._load (module.js:312:12)\n    at Function.Module.runMain (module.js:497:10)\n    at startup (node.js:119:16)\n    at node.js:906:3","code":601,"status":500,"name":"NativeError","message":"Fake native error"}
------------
{"stack":"TypeError: bad type\n    at Object.<anonymous> (/tmp/extractError.js:7:10)\n    at Module._compile (module.js:456:26)\n    at Object.Module._extensions..js (module.js:474:10)\n    at Module.load (module.js:356:32)\n    at Function.Module._load (module.js:312:12)\n    at Function.Module.runMain (module.js:497:10)\n    at startup (node.js:119:16)\n    at node.js:906:3","message":"bad type","name":"TypeError"}
------------
{"stack":"Error: Fake native error\n    at Object.<anonymous> (/tmp/extractError.js:9:19)\n    at Module._compile (module.js:456:26)\n    at Object.Module._extensions..js (module.js:474:10)\n    at Module.load (module.js:356:32)\n    at Function.Module._load (module.js:312:12)\n    at Function.Module.runMain (module.js:497:10)\n    at startup (node.js:119:16)\n    at node.js:906:3","code":601,"status":500,"name":"NativeError","message":"An unexpected NativeError occurred.","cause":{"stack":"TypeError: bad type\n    at Object.<anonymous> (/tmp/extractError.js:7:10)\n    at Module._compile (module.js:456:26)\n    at Object.Module._extensions..js (module.js:474:10)\n    at Module.load (module.js:356:32)\n    at Function.Module._load (module.js:312:12)\n    at Function.Module.runMain (module.js:497:10)\n    at startup (node.js:119:16)\n    at node.js:906:3","message":"bad type","name":"TypeError"}}
```

If possible can you try cloning the branch and give it a spin?

**Comment by bodenr** [comment]
merged and published errors v 0.3.0 which includes the functionality described above.

if you need changes / function you can submit a PR.
